adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -80,33 +86,60 @@ TokenCacheItem getATFromCache(final String resource, final String clientId, fina
         
         return accessTokenItem;
     }
-    
+
     /**
      * @return {@link TokenCacheItem} for regular token cache entry.  
      */
     TokenCacheItem getRegularRefreshTokenCacheItem(final String resource, final String clientId, final String user) {
+        final CacheEvent cacheEvent = startCacheTelemetryRequest(EventStrings.TOKEN_TYPE_RT);
+
         final String cacheKey = CacheKey.createCacheKeyForRTEntry(mAuthority, resource, clientId, user);
-        return mTokenCacheStore.getItem(cacheKey);
+        final TokenCacheItem item =  mTokenCacheStore.getItem(cacheKey);
+
+        if (item != null) {
+            cacheEvent.setTokenTypeRT(true);
+        }
+        Telemetry.getInstance().stopEvent(mTelemetryRequestId, cacheEvent, EventStrings.TOKEN_CACHE_LOOKUP);
+
+        return item;
     }
     
     /**
      * @return {@link TokenCacheItem} for MRRT token cache entry.  
      */
     TokenCacheItem getMRRTItem(final String clientId, final String user) {
+        final CacheEvent cacheEvent = startCacheTelemetryRequest(EventStrings.TOKEN_TYPE_MRRT);
         final String cacheKey = CacheKey.createCacheKeyForMRRT(mAuthority, clientId, user);
-        return mTokenCacheStore.getItem(cacheKey);
+        final TokenCacheItem item = mTokenCacheStore.getItem(cacheKey);
+
+        if (item != null) {
+            cacheEvent.setTokenTypeMRRT(true);
+            cacheEvent.setTokenTypeFRT(item.isFamilyToken());
+        }
+        Telemetry.getInstance().stopEvent(mTelemetryRequestId, cacheEvent, EventStrings.TOKEN_CACHE_LOOKUP);
+
+        return item;
     }
     
     /**
      * @return {@link TokenCacheItem} for FRT token cache entry.  
      */
     TokenCacheItem getFRTItem(final String familyClientId, final String user) {
+        final CacheEvent cacheEvent = startCacheTelemetryRequest(EventStrings.TOKEN_TYPE_FRT);
```

Comment:
Not sure why you think it will be noisy. Also in aggregated mode you will not get everything that's collected.

---
In reply to: [90996130](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/736#discussion_r90996130) [](ancestors = 90996130)